adoptingerlang.org

```diff
@@ -1334,6 +1334,370 @@ Repeat this exercise from time to time, and eventually do some [[https://princip
+   │  ├─ chatterbox─0.9.1 (hex package)
+   │  │  └─ hpack─0.2.3 (hex package)
+   │  ├─ ctx─0.5.0 (hex package)
+   │  └─ gproc─0.8.0 (hex package)
+   ├─ jsx─2.10.0 (hex package)
+   ├─ pgo─0.8.0+build.91.refaf02392 (git repo)
+   │  ├─ backoff─1.1.6 (hex package)
+   │  └─ pg_types─0.0.0+build.24.ref32ed140 (git repo)
+   ├─ recon─2.4.0 (hex package)
+   └─ uuid─1.7.5 (hex package)
+      └─ quickrand─1.7.5 (hex package)
+#+END_SRC
+
+In this listing, all top-level applications (those we write in the repository) are at the top, and the dependencies we just fetched are all below them. You can see the entire resolution tree that way, find which versions have been fetched, and by which application they have been brought in. This can prove useful to understand why a version was selected when a transitive dependency is imported by two or more applications.
+
+If you take a look within =_build/default/lib=, you will see all these applications within their own directory:
+
+#+NAME: service_discovery.deps-ls
+#+BEGIN_SRC sh
+$ ls _build/default/lib
+acceptor_pool  gproc       rfc3339
+backoff        grpcbox     service_discovery
+base32         hpack       service_discovery_grpc
+chatterbox     iso8601     service_discovery_http
+counters       jsx         service_discovery_postgres
+ctx            opencensus  service_discovery_storage
+dns            pgo         telemetry
+elli           pg_types    uuid
+eql            quickrand   wts
+erldns         recon
+#+END_SRC
+
+Each of these is an OTP application with similar directory structures. This layout is rather similar to the project structure described for releases in [[OTP at a High Level]], but this is still just a staging area. If we want to run an actual build, we should aim for releases.
+
+**** Building a Project with Dependencies
+
+Building actual projects requires more than just fetching its dependencies and compiling them. As mentioned in [[What Makes a Lib an App]], the Erlang run-time system expects to find run-time definitions of dependencies within the =.app= file. Not putting them there means that while your libraries are available on disk, they won't necessarily be visible during some build phases, some tests, or in releases.
+
+For everything to work, we must add them. Open up =apps/service_discovery/src/service_discovery.app.src= and look at the values in the =applications= tuple:
+
+#+NAME: service_discovery.app-src
+#+BEGIN_SRC erlang
+{application, service_discovery,
+ [{description, "Core functionality for service discovery service"},
+  {vsn, {git, short}},
+  {registered, []},
+  {mod, {service_discovery_app, []}},
+  {applications,
+   [kernel,
+    stdlib,
+    erldns,
+    service_discovery_storage
+   ]},
+  {env,[]},
+  {modules, []},
+
+  {licenses, ["Apache 2.0"]},
+  {links, []}
+ ]}.
+#+END_SRC
+
+You can see =erldns= and =service_discovery_storage= have been added. Only when such dependencies are specified will everything both build and run fine.
```

Comment:
They may not build fine since I think we currently rely on these tuples for the `-include_lib(...)` directives, breaking the distinction between buildtime and runtime.